src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,77 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private List<Product> products;
+    private long maxId;
+    private Object obj = new Object();
+
+    {
+        this.products = new ArrayList();
+        getSampleProducts();
+
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws ProductNotFoundExeption{
+        synchronized (obj){
+            return products.stream()
+                    .filter(product -> id.equals(product.getId()))
+                    .findAny()
+                    .orElseThrow(() ->new ProductNotFoundExeption());
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        synchronized (obj){
+            return products.stream()
+                    .filter(product -> product.getPrice() != null)
+                    .filter(product -> product.getStock() > 0)
+                    .collect(Collectors.toList());}
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        synchronized (obj){
+            if(product.getId() == null){
+                product.setId(maxId++);
+            }
+                products.add(product);
+        }
     }
 
     @Override
     public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+        synchronized (obj){
+            products = products.stream()
```

Comment:
Bad approach to remove objects like this. Just imagine 10000 element collection, think about memory usage and performance of creating new collection. Better to use designed methods like remove(). I'd use get Product you implemented earlier and remove() method of collection class.